aten/src/ATen/core/LegacyTypeDispatch.h

```diff
@@ -46,14 +65,83 @@ struct AT_CORE_API LegacyTypeDeleter {
 class AT_CORE_API LegacyTypeDispatch {
 public:
   using TypeUniquePtr = std::unique_ptr<Type, LegacyTypeDeleter>;
+  // WARNING: This function has the precondition that you have
+  // initialized the type you want to call.  This initialization
+  // step is generally done by Context, or assumed because you
+  // have a Tensor and thus the Type of that Tensor must already
+  // be initialized.
   Type* getNonVariableTypeRaw(Backend p, ScalarType s) {
     return type_registry[static_cast<int>(p)][static_cast<int>(s)].get();
   }
+  Type * getNonVariableTypeOpt(Backend p, ScalarType s) {
+    if (p != Backend::Undefined) {
+      initBackendIfNeeded(backendToDeviceType(p));
+      initComplexIfNeeded(s);
+    }
+    auto type = getNonVariableTypeRaw(p, s);
+
+    if(!type) {
+      // there is only a single Undefined Type.
+      if (p == Backend::Undefined || s == ScalarType::Undefined) {
+        return getNonVariableTypeRaw(Backend::Undefined, ScalarType::Undefined);
+      }
+    }
+
+    return type;
+  }
+
+  Type & getNonVariableType(Backend p, ScalarType s) {
+    auto* type = getNonVariableTypeOpt(p, s);
+    if (!type) AT_ERROR(toString(p), toString(s), "Type is not enabled.");
+    return *type;
+  }
+
+  Type* getTypeRaw(Backend p, ScalarType s, bool is_variable) {
+    auto baseType = getNonVariableTypeRaw(p, s);
+    if (is_variable) {
+      return &detail::getVariableHooks().getVariableTypeFromBaseType(*baseType);
+    } else {
+      return baseType;
+    }
+  }
+  Type & getVariableType(Backend p, ScalarType s) {
+    auto& baseType = getNonVariableType(p, s);
+    return detail::getVariableHooks().getVariableTypeFromBaseType(baseType);
+  }
+  Type & getType(Backend p, ScalarType s, bool is_variable) {
+    if (is_variable) {
+      return getVariableType(p, s);
+    } else {
+      return getNonVariableType(p, s);
+    }
+  }
   void registerType(Backend b, ScalarType s, TypeUniquePtr&& t) {
     type_registry[static_cast<int>(b)][static_cast<int>(s)] = std::move(t);
     detail::getVariableHooks().registerVariableTypeFor(this, b, s);
   }
 private:
+  void initBackendIfNeeded(DeviceType p) {
```

Comment:
what's the idea behind this?

Reply:
LegacyTypeDispatch may not actually have types for CPU or CUDA loaded at the time you try to access them. If they're not loaded (this can only be true the first time we try to initialize them), we have to go request them to actually be loaded. This might involve properly initializing CUDA, for example, which we want to do lazily because it is quite costly.